backtrace-library/src/main/java/backtraceio/library/breadcrumbs/BacktraceComponentListener.java

```diff
@@ -0,0 +1,99 @@
+package backtraceio.library.breadcrumbs;
+
+import android.content.ComponentCallbacks2;
+import android.content.res.Configuration;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import backtraceio.library.models.types.BacktraceBreadcrumbLevel;
+import backtraceio.library.models.types.BacktraceBreadcrumbType;
+
+public class BacktraceComponentListener implements ComponentCallbacks2 {
+
+    private BacktraceBreadcrumbs backtraceBreadcrumbs;
+
+    public BacktraceComponentListener(BacktraceBreadcrumbs backtraceBreadcrumbs)
+    {
+        this.backtraceBreadcrumbs = backtraceBreadcrumbs;
+    }
+
+    private String getMemoryWarningString(final int level)
+    {
+        switch (level)
+        {
+            case ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN:
+                return "TRIM_MEMORY_UI_HIDDEN";
+            case ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE:
+                return "TRIM_MEMORY_RUNNING_MODERATE";
+            case ComponentCallbacks2.TRIM_MEMORY_RUNNING_LOW:
+                return "TRIM_MEMORY_RUNNING_LOW";
+            case ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL:
+                return "TRIM_MEMORY_RUNNING_CRITICAL";
+            case ComponentCallbacks2.TRIM_MEMORY_BACKGROUND:
+                return "TRIM_MEMORY_BACKGROUND";
+            case ComponentCallbacks2.TRIM_MEMORY_MODERATE:
+                return "TRIM_MEMORY_MODERATE";
+            case ComponentCallbacks2.TRIM_MEMORY_COMPLETE:
+                return "TRIM_MEMORY_COMPLETE";
+            default:
+                return "Generic memory warning";
+        }
+    }
+
+    private BacktraceBreadcrumbLevel getMemoryWarningLevel(final int level)
+    {
+        switch (level)
+        {
+            case ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE:
+            case ComponentCallbacks2.TRIM_MEMORY_RUNNING_LOW:
+            case ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL:
+                return BacktraceBreadcrumbLevel.ERROR;
+            case ComponentCallbacks2.TRIM_MEMORY_BACKGROUND:
+            case ComponentCallbacks2.TRIM_MEMORY_MODERATE:
+            case ComponentCallbacks2.TRIM_MEMORY_COMPLETE:
+                return BacktraceBreadcrumbLevel.FATAL;
```

Comment:
why we treat those memory warnings as a fatal? When user switch application to background we will get this event and we will report is in breadcrumbs as a fatal event?

Reply:
You're right, `TRIM_MEMORY_BACKGROUND` can just be error. I treat `TRIM_MEMORY_MODERATE` and `TRIM_MEMORY_COMPLETE` as fatal since it's more likely our process can be killed in those cases.

https://developer.android.com/reference/android/content/ComponentCallbacks2